chain/client/src/sync/epoch.rs

```diff
@@ -514,6 +508,148 @@ impl EpochSync {
+            &last_epoch.final_block_header_in_next_epoch,
+        )?;
+        Ok(())
+    }
+
+    fn verify_current_epoch_data(
+        current_epoch: &EpochSyncProofCurrentEpochData,
+        final_block_header: &BlockHeader,
+    ) -> Result<(), Error> {
+        // Verify first_block_header_in_epoch
+        let first_block_header = &current_epoch.first_block_header_in_epoch;
+        if !verify_hash(
+            *final_block_header.block_merkle_root(),
+            &current_epoch.merkle_proof_for_first_block,
+            *first_block_header.hash(),
+        ) {
+            return Err(Error::InvalidEpochSyncProof(
+                "invalid merkle_proof_for_first_block".to_string(),
+            ));
+        }
+
+        Ok(())
+    }
+
+    fn verify_epoch_sync_data_hash(
+        last_epoch: &EpochSyncProofLastEpochData,
+        last_epoch_block_producers: &Vec<ValidatorStake>,
+        epoch_manager: &dyn EpochManagerAdapter,
+    ) -> Result<(), Error> {
+        // Verify epoch_sync_data_hash matches final_block_header_in_next_epoch.epoch_sync_data_hash
+        let epoch_sync_data_hash = CryptoHash::hash_borsh(&(
+            &last_epoch.first_block_in_epoch,
+            &last_epoch.second_last_block_in_epoch,
+            &last_epoch.last_block_in_epoch,
+            &last_epoch.epoch_info,
+            &last_epoch.next_epoch_info,
+            &last_epoch.next_next_epoch_info,
+        ));
+        let expected_epoch_sync_data_hash = last_epoch
+            .final_block_header_in_next_epoch
+            .epoch_sync_data_hash()
+            .ok_or(Error::InvalidEpochSyncProof("missing epoch_sync_data_hash".to_string()))?;
+        if epoch_sync_data_hash != expected_epoch_sync_data_hash {
+            return Err(Error::InvalidEpochSyncProof("invalid epoch_sync_data_hash".to_string()));
+        }
+
+        // Verify final_block_header_in_next_epoch
+        let approvers_info = last_epoch_block_producers
+            .iter()
+            .map(|validator| (validator.get_approval_stake(false), false))
+            .collect_vec();
+        let block_header = &last_epoch.final_block_header_in_next_epoch;
+        if !epoch_manager.verify_approval_with_approvers_info(
+            block_header.hash(),
+            block_header.height(),
+            block_header.height() + 1,
+            &last_epoch.approvals_for_final_block_in_next_epoch,
+            approvers_info,
+        )? {
+            return Err(Error::InvalidEpochSyncProof(
+                "invalid final_block_header_in_next_epoch".to_string(),
+            ));
+        }
+        Ok(())
+    }
+
+    fn verify_block_producer_handoff(
+        epoch: &EpochSyncProofPastEpochData,
+        prev_epoch_id: &EpochId,
+        prev_epoch_next_bp_hash: &CryptoHash,
+        epoch_manager: &dyn EpochManagerAdapter,
+    ) -> Result<(), Error> {
+        // Verify epoch.block_producers
+        let bp_hash = Chain::compute_bp_hash_from_validator_stakes(
```

Comment:
Nice refactoring for this function. I was worried earlier about the protocol version check in here, but this is good.